0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+if (with_mask == mask) {
+    bool scoped = !cr[1];
+    while (scoped) {
+        cr[1] = measure qr[1];
+        scoped = !cr[1];
+    }
+}
+```
+
+I don't expect the declaration of scoped variables to be valid for execution on hardware immediately, but I think it's a little easier to implement in Qiskit if the possibility is always there.
+
+The control-flow builder usage in this case would desugar into this long-form construction:
+```python
+from qiskit.circuit import QuantumCircuit, QuantumRegister, ClassicalRegister
+from qiskit.circuit.classical import expr, types
+
+qr = QuantumRegister(3)
+cr = ClassicalRegister(3)
+qc = QuantumCircuit(qr, cr)
+mask = qc.add_var("mask", types.Uint(3), expr.value(5))
+qc.h(0)
+qc.cx(0, 1)
+qc.cx(0, 2)
+qc.measure(0, 0)
+qc.measure(1, 1)
+qc.measure(2, 2)
+with_mask = qc.add_var("with_mask", types.Uint(3), expr.bit_and(mask, cr))
+
+true_body = QuantumCircuit([qr[1]], [cr[1]], captures=[mask, with_mask])
+scoped = true_body.add_var("scoped", types.Bool(), expr.logic_not(cr[1]))
+
+while_body = QuantumCircuit([qr[1]], [cr[1]], captures=[scoped])
+while_body.measure(qr[1], cr[1])
+while_body.store(scoped, expr.logic_not(cr[1]))
+
+true_body.while_loop(scoped, while_body, [qr[1]], [cr[1]])
+
+qc.if_test(expr.equals(with_mask, mask), true_body, [qr[1]], [cr[1]])
+```
+
+
+## Detailed Design
+
+### Modifications to `QuantumCircuit`
+
+In addition to the public API surface discussed above, `QuantumCircuit` will need to keep a track of its variables internally.
+The public-access `iter_*` methods are designed to allow `QuantumCircuit` to use private mutable, growable data structures to do this efficiently.
+
+The purpose of having explicit variable capture is such that it remains impossible for a `QuantumCircuit` to be in an invalid state, and to make explicit the requirement to use the same variables.
+We currently have a problem in the control-flow operations where the exact classical registers must be used in inner blocks (if they are to be accessed), and this constrains the data in the `Clbit` binding between the `CircuitInstruction` and the individual blocks' `.clbits` in a way that is not clear at all, and introduces many possibilities for data incoherence.
+Requiring closures to be explicitly stated sidesteps all of this; there is no longer a "binding" step, and the validity of all the operations can be enforced during construction only.
+Holding a `QuantumCircuit` then contains all the information needed to work with the new objects within it.
+
+Of existing circuit methods:
+- `QuantumCircuit.copy_empty_like` would need to expand to include variables in the copy.
+- `QuantumCircuit.from_instructions` needs to gain the same keyword arguments as the `__init__`.
+- `QuantumCircuit.compose` and `QuantumCircuit.tensor` certainly need to have _some_ handling for this (even if it's just erroring if both circuits contain variables), but it's not clear to me how variable collisions should be handled.
```

Comment:
I don't think you want to just error if both circuits contain variables. I think the right behavior depends on whether we think of `compose` as introducing a new scope or as inlining one circuit into another.

Reply:
I see that you take the "inlining" view below, which feels right to me. I am good with the approach you agreed to with @kevinhartman.